### PR TITLE
[libfreerdp] add bounds check, fix 3 security breaches.

### PR DESCRIPTION
**Pull request by starhcq**
In Windows remote run vulnerablillties exe program, to create Microsoft::Windows::RDS::Graphics/rdpgrfx channel, may case Remmina crash.
So, add bounds check in out of range code.

### CODE REVIEW

File: libfreerdp/gdi/gdi.c [hunk 1]
```diff
@@ -1081,6 +1092,13 @@ static BOOL gdi_surface_bits(rdpContext* context, const SURFACE_BITS_COMMAND* cm
 				         cmd->bmp.bitmapDataLength, size);
 				goto out;
 			}
+
+            // Bound check, check that the paste area is valid,
+            // and determine if the picture is in the paste area.
+            const RECTANGLE_16 rect = { (UINT16)cmd->destLeft, (UINT16)cmd->destTop,
```
Comment (akallabeth) [inline libfreerdp/gdi/gdi.c:1098]: something missing?

File: libfreerdp/codec/planar.c [hunk 2]
```diff
@@ -720,6 +720,11 @@ BOOL planar_decompress(BITMAP_PLANAR_CONTEXT* planar, const BYTE* pSrcData, UINT
 
 	srcp = pSrcData;
 
+    if (!pSrcData) {
```
Comment (akallabeth) [inline libfreerdp/codec/planar.c:723]: good point, this check does not hurt.

File: libfreerdp/gdi/gfx.c [hunk 3]
```diff
@@ -1252,6 +1252,18 @@ static UINT gdi_SolidFill(RdpgfxClientContext* context, const RDPGFX_SOLID_FILL_
...
 	{
 		rect = &(solidFill->fillRects[index]);
+
+        /* Bounds check, limit the size of the requested rect
+         * no larger than the surface data buffer. */
+        if (rect->left > surface->width)
+            goto fail;
+        if (rect->top > surface->height)
+            goto fail;
+        if (rect->right > surface->width)
```
Comment (akallabeth) [inline libfreerdp/gdi/gfx.c:1262]: might be easier to do a rectangle intersect?

**Comment by akallabeth** [comment]
@freerdp-bot test please

**Review by akallabeth — APPROVED** [review]

File: libfreerdp/gdi/gdi.c (hunk 1 above)
Comment (starhcq) [reply libfreerdp/gdi/gdi.c]: >  if (!is_rect_valid(&rect, gdi->width, gdi->height)) {
                WLog_ERR(TAG, "cmd->bmp is outside the scope of the gdi area.");
                goto out;
            }
I need to add this code to judge the rect area.

**Comment by starhcq** [comment]
> @starhcq did refactor your commits a bit, thank you for the fixes!

Wow, 

> @starhcq did refactor your commits a bit, thank you for the fixes!


That's really great, thanks for the perfection!

**Comment by starhcq** [comment]
> @starhcq did refactor your commits a bit, thank you for the fixes!

That's really great, thanks for the perfection!

**Pull request merged**